policyengine_us/variables/gov/states/md/tax/income/agi/subtractions/md_two_income_subtraction.py

```diff
@@ -0,0 +1,61 @@
+from policyengine_us.model_api import *
+
+
+class md_two_income_subtraction(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "MD two-income married couple subtraction from AGI"
+    unit = USD
+    definition_period = YEAR
+    reference = "https://www.marylandtaxes.gov/forms/21_forms/Resident_Booklet.pdf#page=16"
+    defined_for = StateCode.MD
+
+    def formula(tax_unit, period, parameters):
+        person = tax_unit.members
+        # compute head and spouse US AGI portions using irs_gross_income shares
+        us_agi = tax_unit("adjusted_gross_income", period)
+        gross_income = person("irs_gross_income", period)
+        is_head = person("is_tax_unit_head", period) == 1
+        is_spouse = person("is_tax_unit_spouse", period) == 1
+        head_gross_income = tax_unit.sum(where(is_head, gross_income, 0))
+        couple_gross_income = tax_unit.sum(
+            where(is_head | is_spouse, gross_income, 0)
+        )
+        head_frac = where(
+            couple_gross_income > 0, head_gross_income / couple_gross_income, 1
+        )
+        head_us_agi = head_frac * us_agi
+        spouse_us_agi = (1 - head_frac) * us_agi
+
+        # compute head and spouse MD AGI additions using ad hoc rule
+        total_additions = tax_unit("md_total_additions", period)
+        head_adds = 0.5 * total_additions
+        spouse_adds = 0.5 * total_additions
+
+        # compute head and spouse MD AGI subtractions
+        p = parameters(period).gov.states.md.tax.income.agi.subtractions
+        head_subs = 0
```

Comment:
Could this work instead, to be flexible to potential future subtractions?
```suggestion
        subs_except_two_income = [x for x in p.sources if x != "md_two_income_subtraction"]
        total_subs_except_two_income = add(tax_unit, period, subs_except_two_income)
        head_subs = 0.5 * total_subs_except_two_income
        spouse_subs = 0.5 * total_subs_except_two_income
```

Reply:
I don't think that will work because each additional subtraction (and there will be more) needs to be split between head and spouse in a potentially different way.
That is the point of the "case statement" (using the `if` and `elif` and `else` statements).

Reply:
My question is when I get to the `else` in the "case statement", can I `raise` a `ValueError`?   Will that work in the Framework?  Getting to the `else` statement is a fatal error that needs to stop the computations.

Reply:
I think `ValueError` would work.

For parsimony and consistency with the logic in this formula for additions, I'd suggest taking the simpler approach using list comprehension. If we can make these imputations more accurate for each addition and subtraction, we can file an issue to do so, but I'd prefer we implement the approach we're taking in the simplest way possible.

Reply:
Here's a more complete comparison of the code segments:

Current:
```python
        head_subs = 0
        spouse_subs = 0
        for subtraction in p.sources:
            if subtraction == "md_two_income_subtraction":
                continue
            if subtraction == "md_dependent_care_subtraction":
                unit_care_amt = tax_unit(subtraction, period)
                head_frac = where(
                    tax_unit("filing_status", period) == "Joint", 0.5, 1.0
                )
                head_subs += head_frac * unit_care_amt
                spouse_subs += (1 - head_frac) * unit_care_amt
            else:
                print("MD ERROR")
```

My proposal:
```python
# Create a list of all elements in p.sources except for "md_two_income_subtraction".
subs_except_two_income = [x for x in p.sources if x != "md_two_income_subtraction"]
# Compute the total value of subtractions, except this one.
total_subs_except_two_income = add(tax_unit, period, subs_except_two_income)
# Split subtractions equally between head and spouse.
head_subs = 0.5 * total_subs_except_two_income
spouse_subs = 0.5 * total_subs_except_two_income
```

Reply:
How is it wrong? AFAICT, my code could theoretically functionally differ from yours in two ways:
1. It counts all subtractions except the two-income one
2. It skips the different treatment for non-joint filers

Re: (1), line 5 of the [worksheet](https://www.marylandtaxes.gov/forms/21_forms/Resident_Booklet.pdf#page=16) points to lines 8-13 of the 502:
<img width="1023" alt="image" src="https://user-images.githubusercontent.com/6076111/208028731-5dbd229d-5cb4-44b5-9588-ed79c147eb2a.png">

[Lines 8-13 of Form 502](https://www.marylandtaxes.gov/forms/21_forms/502.pdf#page=2) shows this considers all subtractions except the two-income one:
<img width="824" alt="image" src="https://user-images.githubusercontent.com/6076111/208028746-e91cf73c-ad85-4d7c-b19a-f00bc8c79a73.png">

So (1) is not a functional difference. I'd expect that the legal code specifies all subtractions except the two-income subtraction as my code does.

Re: (2), the credit does not apply for non-joint filers so that logic is unnecessary.

Am I missing other differences? Happy to chat through it by phone tomorrow if easier.

Reply:
@MaxGhenis, Maybe I'm making the MD AGI subtractions too complicated, but the current code in PR #1548 seems necessary to fix issues #1568 and #1569, which I would like to do after #1548 is merged.

Your premise (`I'd expect that the legal code specifies all subtractions except the two-income subtraction as my code does.`) seems not to be true.  Or am I misreading the forms and instructions?

Reply:
[Here's the legal code for this subtraction](https://govt.westlaw.com/mdc/Document/NF93A7BD2E6C811ECA065A3F5EAA0E5C9?viewType=FullText&originationContext=documenttoc&transitionType=CategoryPageItem&contextData=(sc.Default)) (could you please add it as a reference?):
>(r)(1) In this subsection, “modified Maryland adjusted gross income” means Maryland adjusted gross income determined separately for each spouse on a joint return without regard to the subtraction allowed under this subsection.
(2) For a two-income married couple filing a joint return, the subtraction under subsection (a) of this section includes the lesser of $1,200 or the modified Maryland adjusted gross income of the spouse with the lesser modified Maryland adjusted gross income for the taxable year.

This confirms my expectation that the legal code does not isolate the dependent care subtraction as your code does. Please accept my suggestion or propose a test case that my simpler code fails.

Reply:
I see that Maryland allows subtractions for personal income variables like pensions. In that case I'd suggest one of these:
1. Add a constant dict to this pointing to the person-level subtractions, and processing those separately while splitting the tax-unit-level ones as provided.
2. Create a person-level Maryland AGI variable and have the tax-unit-level one sum that between head and spouse. This would resemble the MO AGI variable.

Maybe another solution is available. But this code should reflect the legal code in referencing the list of all subtractions, and it can avoid a for loop by using one or two list comprehension + `add` statements. So I'd suggest merging this PR with my code and amending it as necessary in a subsequent PR.

Reply:
OK, I've simplified the code along the lines of your suggestion.